_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,463 @@
+
+    testCases.forEach((key, value) -> dynamicTests.add(DynamicTest.dynamicTest("Test " + key + "->" + value, () -> {
+        // exercise phase
+        KafkaUtils.updateConfigurationWithStabilityWait(CLUSTER_NAME, key, value);
+
+        // verify phase
+        assertThat(KafkaUtils.verifyCrDynamicConfiguration(CLUSTER_NAME, key, value), is(true));
+        assertThat(KafkaUtils.verifyPodDynamicConfiguration(KafkaResources.kafkaStatefulSetName(CLUSTER_NAME), key, value), is(true));
+    })));
+    return dynamicTests.iterator();
+}
+```
+
+It is worth mentioning is that our tests occasionally do chain with exercise and verify phases. So that tests can look like:
+
+1. Exercise (modification of component state)
+2. Verify (verification of component state)
+3. Exercise (another modification of component state)
+4. Verify (another verification modification of component state)
+5. and more...
+
+### Teardown phase
+
+Because we have two stacks for storing resources, cluster resources deletion can be easily performed in `@AfterEach` or `@AfterAll` methods.
+Resource lifecycle implementation will ensure that all resources tied to a specific stack will be deleted in the correct order.
+Teardown is triggered in `@AfterAll` of `AbstractST`:
+```
+@AfterAll
+void teardownEnvironmentClass() {
+    if (Environment.SKIP_TEARDOWN == null) {
+        tearDownEnvironmentAfterAll();
+        teardownEnvForOperator();
+    }
+}
+```
+
+so if you want to change teardown from your `@AfterAll`, you must override method `tearDownEnvironmentAfterAll()`:
+```
+@Override
+protected void tearDownEnvironmentAfterAll() {
+    doSomethingYouNeed();
+    super.tearDownEnvironmentAfterAll();
+}
+```
+
+In order to delete all resources from specific `Resources` instance, execute:
+```
+ResourceManager.deleteMethodResources();
+ResourceManager.deleteClassResources();
+```
+
+## Resources
+
+Resources are in our case an abstraction of the `.yaml` definitions of custom resources provided by Strimzi, such as 
+`Kafka`, `KafkaTopic`, `KafkaUser`, `KafkaConnect`, `KafkaMirrorMaker` and so on.
+These custom resources are represented as Java objects. 
+This yaml representation of a `KafkaTopic`
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 1
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+```
+
+can be also be represented as the following Java code:
+
+```
+KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
+// additional changes...
+kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
+    .withNewMetadata()
+        .withName(topicName)
+        .withNamespace(ResourceManager.kubeClient().getNamespace())
+        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
+    .endMetadata()
+    .editSpec()
```

Comment:
Either spec/config is wrong or java code editSpec() - as they do not correlate:
    retention.ms: 7200000
    segment.bytes: 1073741824
vs
        .addToConfig("min.insync.replicas", minIsr)